### ISSUE TITLE
[BDD] Triggers mise à jour date de création

### ISSUE THREAD

**Issue by ch-cbna** [PnX-SI/GeoNature#3294]
J'ai une question concernant les triggers qui existent sur les champs `meta_create_date` des tables `gn_synthese.t_sources`, `gn_meta.t_acquisition_frameworks`, `gn_meta.t_datasets`, `utilisateurs.t_roles` et `gn_synthese.synthese` : 
Quel est leur intérêt ?

Lorsqu'on importe des données avec des dates de création d'origine, nous souhaitons les conserver dans la base GeoNature, hors pour ce faire nous sommes obligés de désactiver ces triggers avant l'import et de les réactiver après.

**Comment by camillemonchicourt** [comment]
Ce sont des dates automatiques de création dans la BDD.
Cela sert à savoir quand une donnée a été créée dans la BDD.
Cela peut notamment être utile aux API quand tu ne veux récupérer que les nouvelles données en les interrogeant par date de création ou de modification.

**Comment by camillemonchicourt** [comment]
Normalement on doit pouvoir faire en sorte que les triggers ne soient déclenchés que si le champs n'est pas rempli explicitement.
Mais cela peut être nécessaire de calculer automatiquement la date d'intégration de la donnée dans la BDD, donc de ne pas récupérer les dates d'une autre BDD.

**Comment by lpofredc** [comment]
Grand sujet, ces 2 champs font-ils partie de la donnée (au sens standard d'occurence et donc de l'historique de la donnée) ou de l'application (avec le temps, je penche de plus en  plus pour cette option). 

Je ne crois pas que ces valeurs (date de création/date de modification) existent dans le standard SINP (cf. https://inpn.mnhn.fr/docs-web/docs/download/421824).

Si on fait le choix de casser ce trigger pour l'intégration de données en base, c'est a réaliser après mure réflexion. A titre d'exemple, GN2PG s'appuie sur ces champs pour le dispositif de mise à jour incrémentale.

**Comment by camillemonchicourt** [comment]
Oui pour moi ce sont des champs et des infos techniques, pas des infos métiers liées à un standard ou autre.

**Comment by jpm-cbna** [comment]
Je pense que la question était plus de savoir pourquoi utilisé des triggers pour cela à la place d'une valeur par défaut  de ces champs définit à `NOW()` ?
L'avantage de l'utilisation de la valeur par défaut c'est qu'il n'y a pas besoin de désactiver des triggers si on fournit une valeur pour le champ lors d'une requête d'insertion...
Est ce qu'il y a une raison bien précise à l'utilisation des triggers ?

**Comment by lpofredc** [comment]
Les valeurs par défaut y sont déjà. Par contre, si tu fais un `INSERT` avec une valeur `NULL`, elle écrase la valeur `NOW()` définie par défaut, ce qui risque de casser certains fonctionnements applicatifs. Il faudrait peut-être juste revoir le trigger pour appliquer cette règle (extrait dessous) déjà présente dans l'`UPDATE` dans le `CREATE`, pour être certain de ne pas avoir de  valeur `NULL`. Mais aussi d'ajouter des contraintes `NOT NULL` à ces champs?

https://github.com/PnX-SI/GeoNature/blob/30c27266495b4affc635f79748c9984feb81a6d7/backend/geonature/migrations/data/core/public.sql#L25-L26

**Comment by jpm-cbna** [comment]
> Mais aussi d'ajouter des contraintes `NOT NULL` à ces champs?

Justement, il me semble qu'en mettant les champs `NOT NULL` et avec une valeur par défaut à `NOW()` cela va forcer l'utilisation de la fonction `NOW()`. Si on essaye de mettre une valeur NULL dans le champs ou s'il n'est pas indiqué dans l'INSERT, cela devrait exécuter la fonction.

@ch-cbna va le tester. Si cela peut éviter l'utilisation de triggers, c'est surement mieux pour les performances et cela simplifie la maintenance de la base...

**Comment by ch-cbna** [comment]
Effectivement, je confirme, en mettant les champs NOT NULL et avec une valeur par défaut à NOW() cela va forcer l'utilisation de la fonction NOW() si on envoie une valeur NULL.

**Comment by lpofredc** [comment]
@ch-cbna, les tests ont été fait sur les tables avec les triggers ? Car à ma connaissance, Postgresql ne gère pas nativement ce cas de figure, sans les triggers. Si tu cherches à insérer une valeur NULL dans un champ avec une contrainte NOT NULL, il te retourne une erreur pour non respect de la contrainte. 

```psql
CREATE table testts(id int, ts_nullable timestamp default now(), ts_not_nullable timestamp default now() not null);

insert into testts(id, ts_nullable, ts_not_nullable) values (1,null, null);
```

```text
ERROR:  null value in column "ts_not_nullable" of relation "testts" violates not-null constraint
DÉTAIL : Failing row contains (1, null, null).
```

**Comment by ch-cbna** [comment]
J'ai fait le test manuellement dans Dbeaver avec la table `gn_synthese.t_sources`, en activant la contrainte NOT NULL et en supprimant le trigger `tri_meta_dates_t_sources`, puis en lançant la requête : 
```
INSERT INTO gn_synthese.t_sources (
	name_source
)
VALUES (
	'test'
)
```
Je n'ai pas eu d'erreur

**Comment by lpofredc** [comment]
Oui Oui, dans cet exemple, ça marche, parce qu'on n'insère aucune valeur dans ces champs  `meta_{create,update}_date`. 

Mais si tu fais un `INSERT` ou un `UPDATE` avec les champs `meta_{create,update}_date` et que la valeur de ces champs est `NULL`. ce qui peut arriver si tu cherches à importer ces valeurs depuis les données externes, alors tu auras une erreur. Pour moi, le risque est juste là.

**Comment by jpm-cbna** [comment]
Sinon, l'autre solution pour éviter trop de changement, c'est comme tu proposes @lpofredc de faire évoluer la fonction pour forcer l'utilisateur de `NOW()` uniquement si les champs sont `NULL` ou `vide` :
```
CREATE OR REPLACE FUNCTION public.fct_trg_meta_dates_change()
 RETURNS trigger
 LANGUAGE plpgsql
AS $function$
BEGIN
        IF(TG_OP = 'INSERT') THEN
                -- Test if NEW.meta_create_date is null or empty
                if ((NEW.meta_create_date = '') IS NOT FALSE) THEN
                        NEW.meta_create_date = NOW();
                END IF;
        ELSIF(TG_OP = 'UPDATE') THEN
                -- Test if NEW.meta_create_date is null or empty
                if ((NEW.meta_create_date = '') IS NOT FALSE) THEN
                    NEW.meta_update_date = NOW();
                END IF;
                
                -- Test if NEW.meta_update_date is null or empty
                if ((NEW.meta_update_date = '') IS NOT FALSE) THEN
                        NEW.meta_create_date = NOW();
                END IF;
        END IF;
        return NEW;
END;
$function$
;
```
Pour l'expression permettant de tester `NULL` ou `vide`, voir [cette réponse Stackoverflow](https://stackoverflow.com/a/23767625/13311850).
